### PR TITLE
Fix #19655 (TypeError on empty stacktrace)

### LINKED ISSUE #19655: TypeError on empty stacktrace
```
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.6.0-dev.1618 (2016-12-19 18:51 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit dce82fcbc* (0 days old master)
|__/                   |  x86_64-apple-darwin16.3.0

julia> catch_stacktrace()
ERROR: TypeError: stacktrace: in typeassert, expected Array{StackFrame,1}, got Array{Any,1}
 in stacktrace(::Array{Ptr{Void},1}, ::Bool) at ./stacktraces.jl:147
 in catch_stacktrace() at ./stacktraces.jl:166
```

Also happens on 0.5.

One-line PR fix incoming!

### PR DESCRIPTION
**Pull request by iamed2**
The issue was that `vcat(StackFrame[]...)` gives `Any[]`, not `StackFrame[]`.

### CODE REVIEW

**Comment by iamed2** [comment]
This should probably be backported

File: base/stacktraces.jl [hunk 1]
```diff
@@ -144,7 +144,7 @@ doesn't return C functions, but this can be enabled.) When called without specif
 trace, `stacktrace` first calls `backtrace`.
 """
 function stacktrace(trace::Vector{Ptr{Void}}, c_funcs::Bool=false)
-    stack = vcat(map(lookup, trace)...)::StackTrace
+    stack = vcat(StackTrace(), map(lookup, trace)...)::StackTrace
```
Comment (vtjnash) [inline base/stacktraces.jl]: Would: `StackTrace[ lookup(x) for x in trace ]` make more sense?

Comment (iamed2) [inline base/stacktraces.jl]: That wouldn't concatenate the results of `lookup` together.

**Comment by iamed2** [comment]
AppVeyor failure seems like an unrelated network issue.

**Comment by StefanKarpinski** [comment]
Bump.

**Comment by iamed2** [comment]
Wrote a test but setting up a new computer so I have to wait for a Julia build.

**Comment by iamed2** [comment]
I don't understand what's happening with the CI here.

**Comment by tkelman** [comment]
doc build failure from https://github.com/JuliaLang/julia/pull/19268

**Comment by tkelman** [comment]
Ah but that was only on Travis, the AppVeyor failures look like the newly-added tests didn't work there.

**Comment by iamed2** [comment]
Dang; I decided to use `Base.runtests` instead of `make test` for the first time and forgot to rebuild before testing locally.

File: test/stacktraces.jl [hunk 2]
```diff
@@ -124,3 +124,12 @@ let ctestptr = cglobal((:ctest, "libccalltest")),
     @test ctest[1].from_c
     @test ctest[1].pointer === UInt64(ctestptr)
 end
+
+# #19655
+let
+    # not in a `catch`, so should return an empty StackTrace
+    st = stacktrace(empty!(backtrace()))
+    println(st)
```
Comment (tkelman) [inline test/stacktraces.jl]: should probably take this part out?

Comment (iamed2) [inline test/stacktraces.jl]: Dammit. Sorry. 

No need to print and check, the tests below that check for all relevant information.

**Comment by tkelman** [comment]
this is getting overly familiar: appveyor failure was #16555, travis 32 bit linux failure was #19803

**Pull request merged**